File: go-code/encapsulation/encapsulate_collection/after/Person.go
package before

type Person struct {
	Name    string
	courses []Course
}

func (p Person) addCourse(course Course) {
	p.courses = append(p.courses, course)
}

func (p Person) removeCourse(course Course) {

	index := 0
	for i, c := range p.courses {
		if c == course {
			index = i
			break
		}
	}

	p.courses = append(p.courses[:index], p.courses[index+1:]...)
}

func (p Person) getCourses() []Course {
	copyOfCourses := make([]Course, len(p.courses))

	copy(copyOfCourses, p.courses)

	return copyOfCourses
}

File: java-code/src/main/java/refactoring/encapsulation/encapsulate_collection/after/Person.java
package refactoring.encapsulation.encapsulate_collection.after;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import lombok.Getter;
import lombok.Setter;


public class Person {
  @Getter
  @Setter
  private String name;

  private List<Course> courseList = new ArrayList<>();

  //Step 1: Encapsulate the list by adding add and remove methods
  public void addCourse(Course course) {
    this.courseList.add(course);
  }

  public void removeCourse(Course course) {
    this.courseList.remove(course);
  }

  public void removeCourseByName(String name) {
    this.courseList.removeIf(c -> c.getName().equals(name));
  }

  //Step 2: Return a copy of courseList to complete the encapsulation
  public List<Course> getCourseList() {
    return Collections.unmodifiableList(this.courseList);
  }
}
